`char16_t` doesn't work

```
gen/content/browser/download/save_package.rs/gen0.h:187:211: error: no matching member function for call to 'ReplaceComponents'
inline std::unique_ptr<GURL> ReplaceComponents1_autocxx_wrapper(const GURL& autocxx_gen_this, const url_StringPieceReplacements_uint16_t_AutocxxConcrete& arg1)  { return std::make_unique<GURL>(autocxx_gen_this.ReplaceComponents(arg1)); }
                                                                                                                                                                                                 ~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~
../../url/gurl.h:169:8: note: candidate function not viable: no known conversion from 'const StringPieceReplacements<uint16_t>' to 'const StringPieceReplacements<char>' for 1st argument
  GURL ReplaceComponents(const Replacements& replacements) const;
       ^
../../url/gurl.h:170:8: note: candidate function not viable: no known conversion from 'const StringPieceReplacements<uint16_t>' to 'const StringPieceReplacements<char16_t>' for 1st argument
  GURL ReplaceComponents(const ReplacementsW& replacements) const;
       ^
1 error generated.
ninja: build stopped: subcommand failed.
```

Simplified test:

```cpp
inline void take(char16_t &) {}
```

bindgen gives us:

```rust
    #[allow(non_snake_case, non_camel_case_types, non_upper_case_globals)]
    pub mod root {
        #[allow(unused_imports)]
        use self::super::root;
        extern "C" {
            #[cpp_semantics(arg_type_reference(arg1))]
            #[link_name = "\u{1}__Z4takeRDs"]
            pub fn take(arg1: *mut u16);
        }
    }
```

which gives

```
gen0.cxx:75:10: error: cannot initialize a variable of type 'void (*)(::std::uint16_t &)' (aka 'void (*)(unsigned short &)') with an lvalue of type 'void (char16_t &)': type mismatch at 1st parameter ('::std::uint16_t &' (aka 'unsigned short &') vs 'char16_t &')
```

--- a/engine/Cargo.toml
+++ b/engine/Cargo.toml
@@ -30,7 +30,7 @@ log = "0.4"
 proc-macro2 = "1.0.11"
 quote = "1.0"
 indoc = "1.0"
-autocxx-bindgen = "=0.59.14"
+autocxx-bindgen = "=0.59.15"
 #autocxx-bindgen = { git = "https://github.com/adetaylor/rust-bindgen", branch = "denote-deleted-move-constructors" }
 itertools = "0.10.3"
 cc = { version = "1.0", optional = true }

--- a/engine/src/known_types.rs
+++ b/engine/src/known_types.rs
@@ -501,6 +501,14 @@ fn create_type_database() -> TypeDatabase {
         false,
         false,
     ));
+    db.insert(TypeDetails::new(
+        "autocxx::c_char16_t",
+        "char16_t",
+        Behavior::CVoid,
+        Some("c_char16_t".into()),
+        false,
+        false,
+    ));
     db
 }
 

--- a/engine/src/known_types.rs
+++ b/engine/src/known_types.rs
@@ -27,6 +27,7 @@ enum Behavior {
     CByValue,
     CVariableLengthByValue,
     CVoid,
+    CChar16,
     RustContainerByValueSafe,
 }
 
@@ -205,6 +206,7 @@ impl TypeDatabase {
                         | Behavior::RustByValue
                         | Behavior::CByValue
                         | Behavior::CVariableLengthByValue
+                        | Behavior::CChar16
                         | Behavior::RustContainerByValueSafe => true,
                         Behavior::CxxString
                         | Behavior::CxxContainerNotByValueSafe
@@ -283,7 +285,7 @@ impl TypeDatabase {
             .map(|td| {
                 matches!(
                     td.behavior,
-                    Behavior::CVariableLengthByValue | Behavior::CVoid
+                    Behavior::CVariableLengthByValue | Behavior::CVoid | Behavior::CChar16
                 )
             })
             .unwrap_or(false)
@@ -504,7 +506,7 @@ fn create_type_database() -> TypeDatabase {
     db.insert(TypeDetails::new(
         "autocxx::c_char16_t",
         "char16_t",
-        Behavior::CVoid,
+        Behavior::CChar16,
         Some("c_char16_t".into()),
         false,
         false,

--- a/engine/src/lib.rs
+++ b/engine/src/lib.rs
@@ -288,6 +288,7 @@ impl IncludeCppEngine {
             .use_specific_virtual_function_receiver(true)
             .cpp_semantic_attributes(true)
             .represent_cxx_operators(true)
+            .use_distinct_char16_t(true)
             .layout_tests(false); // TODO revisit later
         for item in known_types().get_initial_blocklist() {
             builder = builder.blocklist_item(item);

--- a/integration-tests/tests/integration_test.rs
+++ b/integration-tests/tests/integration_test.rs
@@ -6611,23 +6611,9 @@ fn test_extern_rust_fn_in_mod() {
 #[test]
 fn test_issue_956() {
     let hdr = indoc! {"
-        template <typename b> class c : b {};
-        class GURL {
-        public:
-          typedef c<char16_t> d;
-          GURL e();
-          GURL e(d &);
-        };
+        inline void take(char16_t &) {}
     "};
-    run_test_ex(
-        "",
-        hdr,
-        quote! {},
-        directives_from_lists(&["GURL"], &[], None),
-        make_cpp17_adder(),
-        None,
-        None,
-    );
+    run_test("", hdr, quote! {}, &["take"], &[]);
 }
 
 #[test]

--- a/integration-tests/tests/integration_test.rs
+++ b/integration-tests/tests/integration_test.rs
@@ -6609,6 +6609,28 @@ fn test_extern_rust_fn_in_mod() {
 }
 
 #[test]
+fn test_issue_956() {
+    let hdr = indoc! {"
+        template <typename b> class c : b {};
+        class GURL {
+        public:
+          typedef c<char16_t> d;
+          GURL e();
+          GURL e(d &);
+        };
+    "};
+    run_test_ex(
+        "",
+        hdr,
+        quote! {},
+        directives_from_lists(&["GURL"], &[], None),
+        make_cpp17_adder(),
+        None,
+        None,
+    );
+}
+
+#[test]
 fn test_pv_subclass_mut() {
     let hdr = indoc! {"
     #include <cstdint>

--- a/integration-tests/tests/integration_test.rs
+++ b/integration-tests/tests/integration_test.rs
@@ -6611,9 +6611,18 @@ fn test_extern_rust_fn_in_mod() {
 #[test]
 fn test_issue_956() {
     let hdr = indoc! {"
-        inline void take(char16_t &) {}
+        #include <cstdint>
+        inline void take_int(int&) {}
+        inline void take_uin16(uint16_t&) {}
+        inline void take_char16(char16_t &) {}
     "};
-    run_test("", hdr, quote! {}, &["take"], &[]);
+    run_test(
+        "",
+        hdr,
+        quote! {},
+        &["take_int", "take_uin16", "take_char16"],
+        &[],
+    );
 }
 
 #[test]

--- a/src/lib.rs
+++ b/src/lib.rs
@@ -374,6 +374,20 @@ unsafe impl cxx::ExternType for c_void {
     type Kind = cxx::kind::Trivial;
 }
 
+/// A C++ `char16_t`
+#[allow(non_camel_case_types)]
+#[repr(transparent)]
+pub struct c_char16_t(pub u16);
+
+/// # Safety
+///
+/// We assert that the namespace and type ID refer to a C++
+/// type which is equivalent to this Rust type.
+unsafe impl cxx::ExternType for c_char16_t {
+    type Id = cxx::type_id!(c_char16_t);
+    type Kind = cxx::kind::Trivial;
+}
+
 /// autocxx couldn't generate these bindings.
 /// If you come across a method, type or function which refers to this type,
 /// it indicates that autocxx couldn't generate that binding. A documentation

